plugin/associations.go

```diff
@@ -0,0 +1,133 @@
+	if hasMany == nil {
+		hasMany = &gorm.HasManyOptions{}
+		opts.HasMany = hasMany
+	}
+	var assocKey *Field
+	var assocKeyName string
+	if assocKeyName = generator.CamelCase(hasMany.GetAssociationForeignkey()); assocKeyName == "" {
+		assocKeyName, assocKey = p.findPrimaryKey(parent)
+	} else {
+		var ok bool
+		assocKey, ok = parent.Fields[assocKeyName]
+		if !ok {
+			p.Fail("Missing", assocKeyName, "field in", parent.Name, ".")
+		}
+	}
+	hasMany.AssociationForeignkey = &assocKeyName
+	foreignKey := &Field{Type: assocKey.Type, GormFieldOptions: &gorm.GormFieldOptions{Tag: hasMany.GetForeignkeyTag()}}
+	var foreignKeyName string
+	if foreignKeyName = hasMany.GetForeignkey(); foreignKeyName == "" {
+		foreignKeyName = fmt.Sprintf(fieldName + typeName + assocKeyName)
+	}
+	hasMany.Foreignkey = &foreignKeyName
+	if exField, ok := child.Fields[foreignKeyName]; !ok {
+		child.Fields[foreignKeyName] = foreignKey
+	} else {
+		if exField.Type != foreignKey.Type {
+			p.Fail("Cannot include", foreignKeyName, "field into", child.Name, "as it already exists there with a different type.")
+		}
+	}
+	var posField string
+	if posField = generator.CamelCase(hasMany.GetPositionField()); posField != "" {
+		if exField, ok := child.Fields[posField]; !ok {
+			child.Fields[posField] = &Field{Type: "int", GormFieldOptions: &gorm.GormFieldOptions{Tag: hasMany.GetPositionFieldTag()}}
+		} else {
+			if strings.Contains(exField.Type, "int") {
+				p.Fail("Cannot include", posField, "field into", child.Name, "as it already exists there with a different type.")
+			}
+		}
+		hasMany.PositionField = &posField
+	}
+}
+
+func (p *OrmPlugin) parseHasOne(typeName string, parent *OrmableType, fieldName string, child *OrmableType, opts *gorm.GormFieldOptions) {
+	hasOne := opts.GetHasOne()
+	if hasOne == nil {
+		hasOne = &gorm.HasOneOptions{}
+		opts.HasOne = hasOne
+	}
+	var assocKey *Field
+	var assocKeyName string
+	if assocKeyName = generator.CamelCase(hasOne.GetAssociationForeignkey()); assocKeyName == "" {
+		assocKeyName, assocKey = p.findPrimaryKey(parent)
+	} else {
+		var ok bool
+		assocKey, ok = parent.Fields[assocKeyName]
+		if !ok {
+			p.Fail("Missing", assocKeyName, "field in", parent.Name, ".")
+		}
+	}
+	hasOne.AssociationForeignkey = &assocKeyName
+	foreignKey := &Field{Type: assocKey.Type, GormFieldOptions: &gorm.GormFieldOptions{Tag: hasOne.GetForeignkeyTag()}}
+	var foreignKeyName string
+	if foreignKeyName = generator.CamelCase(hasOne.GetForeignkey()); foreignKeyName == "" {
+		foreignKeyName = fmt.Sprintf(fieldName + typeName + assocKeyName)
+	}
+	hasOne.Foreignkey = &foreignKeyName
+	if exField, ok := child.Fields[foreignKeyName]; !ok {
+		child.Fields[foreignKeyName] = foreignKey
+	} else {
+		if exField.Type != foreignKey.Type {
+			p.Fail("Cannot include", foreignKeyName, "field into", child.Name, "as it already exists there with a different type.")
+		}
+	}
+}
+
+func (p *OrmPlugin) findPrimaryKey(ormable *OrmableType) (string, *Field) {
```

Comment:
GORM does support compound primary keys. Is this something we want to consider here, or should we leave it at single field keys for now or indefinitely?

Reply:
Well, I also have it in my mind. It seems for me that GORM doesn't support association through more than one field out-of-the-box, i.e. you need to use inline preload options or write custom preloading like descibed [here](http://gorm.io/docs/preload.html#Preload). 
We can actually handle this in generator, but it's a question if this will be used a lot. I would better think about how we can expose the query to a user so one can customize it in a unified way without overriding the entire handler. I know GORM hooks can do a lot for this, but I'm still not confident what actually can be customized, since GORM API is not very well documented. Another way is to write our own hooks like `After/Before-ORM/PB` and expose `*gorm.DB` there. 
Anyway, I think we should thoroughly revisit our customization approach in general since no discussion took place on this and there are a lot of similar delicate questions which need to be addressed.

Reply:
Maybe @johnbelamaric or @dmacthedestroyer have something one this.

Reply:
We don't have a need for composite keys right now. We can put it in the backlog.